llama-index-core/llama_index/core/memory/chat_summary_memory_buffer.py

```diff
@@ -0,0 +1,245 @@
+        """Get chat history."""
+        chat_history = self.get_all()
+        if len(chat_history) == 0:
+            return []
+
+        # Give the user the choice whether to count the system prompt or not
+        if self.count_initial_tokens:
+            if initial_token_count > self.token_limit_full_text:
+                raise ValueError("Initial token count exceeds token limit")
+            self._token_count = initial_token_count
+
+        (
+            chat_history_full_text,
+            chat_history_to_be_summarized,
+        ) = self._split_messages_summary_or_full_text(chat_history)
+
+        if self.summarizer_llm is None or len(chat_history_to_be_summarized) == 0:
+            # Simply remove the message that don't fit the buffer anymore
+            updated_history = chat_history_full_text
+        else:
+            updated_history = [
+                self._summarize_oldest_chat_history(chat_history_to_be_summarized),
+                *chat_history_full_text,
+            ]
+
+        self.reset()
+        self._token_count = 0
+        self.set(updated_history)
+
+        return updated_history
+
+    def get_all(self) -> List[ChatMessage]:
+        """Get all chat history."""
+        return self.chat_store.get_messages(self.chat_store_key)
+
+    def put(self, message: ChatMessage) -> None:
+        """Put chat history."""
+        # ensure everything is serialized
+        self.chat_store.add_message(self.chat_store_key, message)
+
+    def set(self, messages: List[ChatMessage]) -> None:
+        """Set chat history."""
+        self.chat_store.set_messages(self.chat_store_key, messages)
+
+    def reset(self) -> None:
+        """Reset chat history."""
+        self.chat_store.delete_messages(self.chat_store_key)
+
+    def get_token_count(self) -> int:
+        """Returns the token count of the memory buffer (excluding the last assistant response)."""
+        return self._token_count
+
+    def _token_count_for_messages(self, messages: List[ChatMessage]) -> int:
+        """Get token count for list of messages."""
+        if len(messages) <= 0:
+            return 0
+
+        msg_str = " ".join(str(m.content) for m in messages)
+        return len(self.tokenizer_fn(msg_str))
+
+    def _split_messages_summary_or_full_text(
+        self, chat_history: List[ChatMessage]
+    ) -> (List[ChatMessage], List[ChatMessage]):
+        """Determine which messages will be included as full text,
+        and which will have to be summarized by the summarizer_llm.
+        """
+        chat_history_full_text = []
+        message_count = len(chat_history)
+        while (
+            message_count > 0
+            and self._token_count + self._token_count_for_messages([chat_history[-1]])
+            <= self.token_limit_full_text
+        ):
+            # traverse the history in reverse order, when token limit is about to be
+            # exceeded, we stop, so remaining messages are summarized

```

Comment:
we need to add some special handling for TOOL and ASSISTANT messages -- see the `get()` method for the `ChatMemoryBuffer` 
- we can't start the chat history with an assistant message (this breaks several LLM APIs)
- we need to maintain the TOOL + ASSISTANT message pairs, so as not not break openai, mistral, etc.

Reply:
I implemented the changes you suggested. I have a few remarks. It may not make sense to summarize tool/function messages, so it would probably be helpful to think about what should happen with such messages in general (completely removed, include as full text, summarize, make it configurable for the user?). Perhaps this is sufficient for now and we can expand on it later.

Also, when a summary is generated (LLM is not none and the messages do not fit in the buffer), the first message will always be a SYSTEM message, followed by potentially an ASSISTANT message. Does that also fail those API’s you mentioned? If so, we need this handling when a summary is made, which the current implementation does. However, if SYSTEM followed by ASSISTANT does not break the API’s, the special handling could be limited to the case when a summary is not generated (when LLM is None or the buffer is not full).